subsys/net/ip/net_tls.c

```diff
@@ -57,3 +57,57 @@ int net_tls_enable(struct net_context *context, bool enabled)
 
 	return 0;
 }
+
+int net_tls_connect(struct net_context *context, bool listening)
+{
+	enum net_context_state state;
+	int err, role, type;
+
+	if (!context) {
+		return -EINVAL;
+	}
+
+	if (!context->options.tls) {
+		return 0;
+	}
+
+	state = net_context_get_state(context);
+	if (state != NET_CONTEXT_IDLE && state != NET_CONTEXT_UNCONNECTED) {
+		return -EBUSY;
+	}
+
+	role = listening? MBEDTLS_SSL_IS_SERVER: MBEDTLS_SSL_IS_CLIENT;
+
+	type = (net_context_get_type(context) == SOCK_STREAM)?
+		MBEDTLS_SSL_TRANSPORT_STREAM:
+		MBEDTLS_SSL_TRANSPORT_DATAGRAM;
+
+	err = mbedtls_ssl_config_defaults(&context->mbedtls.config, role, type,
```

Comment:
Hi @pfl,

I've been trying to come up to speed on the two approaches since they came up at the TSC last week. I've read as much as I could, but would appreciate some clarifications.

In response to @pfalcon's original zstream PR, @tbursztyka said:

> The problem is that it's taking the problem by the easiest part: read, write, flush and close.
But TLS will needs way more than that, and dedicated stuff such as configure and handshake for instance, are annoying parts. How is this approach going to address such requirements?

I am not really clear on how these concerns are addressed by the setsockopt() approach, either.

Here the default SSL configuration is chosen, and the handshake below is performed in the same routine. I'm not sure how an application programmer would set up root certificates etc. for authenticating the host, or set up their own certificates, or configure cipher suites, or manage protocol extensions like SNI, within this API. 

In particular, is there a design written down anywhere describing the setsockopt() plan, similarly to issues #6086, #5900 for zstream?

So far I have read a lot of praise for the TI SimpleLink API approach. I haven't read that API, so perhaps the answer to all of these questions is "we'll do it like SimpleLink does".

But I'm assuming given all your work to integrate with Zephyr's existing sockets that your approach is substantially more complex in the details than just adopting those APIs wholesale.

Any clarification appreciated!

Reply:
> Here the default SSL configuration is chosen, and the handshake below is performed in the same routine. I'm not sure how an application programmer would set up root certificates etc. for authenticating the host, or set up their own certificates, or configure cipher suites, or manage protocol extensions like SNI, within this API.

In this example (compile time) defaults are used for everything. rlubos above showed more options above in line with what a non-trivial application might want to do.

Reply:
> In this example (compile time) defaults are used for everything. rlubos above showed more options above in line with what a non-trivial application might want to do.

Sure, and I saw that.

Are you saying you are going to adopt his proposal as-is?

I'm still not understanding what the design will be which can address the concerns originally outlined by @tbursztyka, and which seem reasonable to me.

From the public code on GitHub and posts to the mailing lists, what is visible so far looks like iterative design by development. My question is if you do have a design document somewhere, or if this is in fact what is going on.

Reply:
I can talk with tbursztyka as I don't have his comments in front of me, nor can I remember what they were about. This PR is WIP and iterative until I have figured out what is needed. rlubos comments will be wrapped into a patch once we're that far. But first I'd like to iterate over what the most default case should be doing.